File: explicit-java/src/main/kotlin/explicit/token/Alias.kt
package explicit.token

import explicit.api.IToken
import explicit.api.Navigator
import explicit.evaluator.ExplicitRuleEvaluation
import explicit.parser.ExplicitRules
import java.util.*

/**
 * Token implementation which represents a token with an explicit alias.
 */
data class Alias(val alias: String, val token: IToken) : IToken {
    override fun asText() = alias
    override fun toString() = "$token:$alias"

    override fun evaluate(rulez: ExplicitRules, navigator: Navigator<String>): ExplicitRuleEvaluation {
        val evaluation = token.evaluate(rulez, navigator)
        val entries = HashMap<String, String>()
        if (evaluation.result)
            entries[alias] = rulez.mappings[navigator.getCurr()] ?: navigator.getCurr()
        return ExplicitRuleEvaluation(evaluation.result, entries, listOf(navigator.getIndex() - 1))
    }

}

File: explicit-python/explicit_nlu/token/Alias.py
import logging
from dataclasses import dataclass

from explicit_nlu.api import IToken, Navigator
from explicit_nlu.evaluator import ExplicitEvaluation
from explicit_nlu.parser import ExplicitRules


#
# Token implementation which represents a token with an explicit alias.
#
@dataclass
class Alias(IToken):
    alias: str
    token: IToken

    def __str__(self):
        return str(self.token) + ":" + self.alias

    def __repr__(self):
        return self.__str__()

    def as_text(self):
        return self.alias

    def evaluate(self, rulez: ExplicitRules, navigator: Navigator) -> ExplicitEvaluation:
        logging.debug("evaluate alias")

        evaluation = self.token.evaluate(rulez, navigator)
        entries = dict()

        if evaluation.result:
            curr = navigator.get_curr()
            entries[self.alias] = rulez.mappings[curr] if curr in rulez.mappings else curr

        return ExplicitEvaluation(evaluation.result, entries, [navigator.get_absolute_index() - 1])
